compiler_opt/tools/bc_hyperparam_tuning.py

```diff
@@ -0,0 +1,92 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script for testing different hyperparameter combinations for BC training."""
+
+import os
+
+from absl import app
+from absl import flags
+import tempfile
+import gin
+
+from compiler_opt.rl import train_bc
+from compiler_opt.tools import generate_default_trace
+
+import tensorflow as tf
+from tensorboard.plugins.hparams import api as hp
+
+_DATA_PATH = flags.DEFINE_multi_string(
+    'data_path', [],
+    'Path to TFRecord file(s) containing training data. Skip training and dump'
+    'an untrained model with random weights (for testing purpose) if '
+    'unspecified.')
+_CORPUS_PATH = flags.DEFINE_string(
+    'corpus_path', '', 'Path to the corpus to evaluate the models.')
+_OUTPUT_DIR = flags.DEFINE_string('output_dir', '', 'Path to the output data')
+_GIN_FILES = flags.DEFINE_multi_string(
+    'gin_files', [], 'List of paths to gin configuration files.')
+_GIN_BINDINGS = flags.DEFINE_multi_string(
+    'gin_bindings', [],
+    'Gin bindings to override the values set in the config files.')
+
+METRIC_REWARD = 'reward'
+
+
+def run_training():
+  with tempfile.TemporaryDirectory() as temp_dir:
+    root_dir = os.path.expanduser(temp_dir)
+    root_dir = os.path.normpath(root_dir)
+    # TODO(boomanaiden154): leave num_iterations and batch_size to config
+    train_bc.train_eval(
+        num_iterations=100,
+        batch_size=64,
+        root_dir=root_dir,
+        data_path=_DATA_PATH.value)
+    return generate_default_trace.generate_default_trace(
+        data_path=_CORPUS_PATH.value,
+        policy_path=os.path.join(root_dir, './saved_policy'),
+        sampling_rate=0.01)
+
+
+def run_tuning():
+  hp_num_units_layer1 = hp.HParam('num_units_layer1', hp.Discrete([40, 80]))
```

Comment:
what's the extra functionality this hyper-parameter tuning can do? is it equivalent to a simple shell script?

Reply:
It does quite a bit more than what a shell script could. It ties in directly to tensorboard through the HParams API and it can be setup to do much more complex searching through the possible space of hyperparameters than brute force like it is set up to do currently (although that isn't implemented at all yet).

Reply:
I checked the documentation here https://www.tensorflow.org/tensorboard/hyperparameter_tuning_with_hparams a bit. I don't see much extra functionality with this tool? e.g., smarter tuning algorithms.

the integration to tensorboard is not so significant actually

Reply:
by looking at the code https://github.com/tensorflow/tensorboard/tree/78db135e980fff6852cd91f3da439b45e7f9fc1e/tensorboard/plugins/hparams

it seems that it is not something under active development, so i'm wondering whether it is the best tool to use?

Reply:
That github link that you sent is to a commit that is pretty old. Looking at head-of-tree, it definitely seems like it isn't under active development, but it at least seems to be getting some maintenance (last commit in December of last year). 

In regards to your other comment, you're right in that the tensorboard integration specifically doesn't allow for better tuning algorithms. That would have to be implemented on our side, but it is an example of something that couldn't easily be done with a shell script. The integration isn't anything too significant, but I thought the visualization method that they use looked fairly useful for actually deciding what hyperparameters are important and what values are better. However, I'm not too familiar with any other options for tuning.

Reply:
i'm not familiar with open source smart tuning packages as well, you can try to search for "bayesian optimization hyperparameter tuning" to see whether there are any good open source ones

Reply:
Switched over to using `keras_tuner` which fits in well with a lot of our existing tooling. Using the bayesian optimizer available through that package.